rfc/system/5606-json-idl.md

```diff
@@ -0,0 +1,343 @@
+- `dict` uses `msgpack.dumps` to turn a dict value to a JSON byte string, and store is to protobuf Json .
+
+###### Convert Literal to Python Value
+
+`Dict[type, type]` uses type hints to convert LiteralMap to Python Value.
+
+`dict` uses `msgpack.loads` to turn a JSON byte string to a `dict` value and store it to protobuf `Struct` .
+
+#### Dataclass Transformer
+##### Before
+###### Convert Python Value to Literal
+Uses `mashumaro JSON Encoder` to turn a dataclass value to a JSON string, and store it to protobuf `Struct` .
+Note: For `FlyteTypes`, we will inherit mashumaro `SerializableType` to define our own serialization behavior, which includes upload file to remote storage.
+
+###### Convert Literal to Python Value
+Uses `mashumaro JSON Decoder` to turn a JSON string to a python value, and recursively fixed int attributes to int (it will be float because we stored it in to `Struct`).
+
+Note: For `FlyteTypes`, we will inherit the mashumaro `SerializableType` to define our own serialization behavior, which includes uploading files to remote storage.
+##### After
+###### Convert Python Value to Literal
+Uses `msgpack.dumps()` to turn a Python value into a byte string.
+Note: For `FlyteTypes`, we will need to customize serialization behavior by msgpack reference here.
+
+https://github.com/msgpack/msgpack-python?tab=readme-ov-file#packingunpacking-of-custom-data-type 
+
+###### Convert Literal to Python Value
+
+Uses `msgpack.loads()` to turn a byte string into a Python value.
+
+Note: For `FlyteTypes`, we will need to customize deserialization behavior by `msgpack` reference here.
+
+https://github.com/msgpack/msgpack-python?tab=readme-ov-file#packingunpacking-of-custom-data-type 
+
+
+#### Pydantic Transformer
+##### Before
+###### Convert Python Value to Literal
+Convert `BaseModel` to a JSON string, and then convert it to a Protobuf `Struct`.
+###### Convert Literal to Python Value
+Convert Protobuf `Struct` to a JSON string and then convert it to a `BaseModel`.
+##### After
+###### Convert Python Value to Literal
+Convert the Pydantic `BaseModel` to a JSON string, then convert the JSON string to a `dictionary`, and finally, convert it to a `byte string` using msgpack.
+###### Convert Literal to Python Value
+Convert `byte string` to a `dictionary` using `msgpack`, then convert dictionary to a JSON string, and finally, convert it to Pydantic `BaseModel`.
+
+### FlyteCtl
+In Flytectl, we can construct input for the execution, so we have to make sure the values we passed to FlyteAdmin can all be constructed to Literal.
+
+https://github.com/flyteorg/flytectl/blob/131d6a20c7db601ca9156b8d43d243bc88669829/cmd/create/serialization_utils.go#L48 
+
+### FlyteConsole
+#### Show input/output on FlyteConsole
+We will get node’s input output literal value by FlyteAdmin’s API, and get the json byte string in the literal value.
+
+We can use MsgPack dumps the json byte string to a dictionary, and shows it to the flyteconsole.
```

Comment:
To make sure I understand correctly, the user will still be able to see the content of the dict, ... because flyte console too will deserialize?
If yes, this is quite neat.

Reply:
Yes, you're right, thank you